File: wasm/engine.c
/**
 * @file WebAssembly transcription of the neural.js inference engine.
 *
 * MEASUREMENT PROTOTYPE - nothing here ships; neural.js stays the
 * production engine. The point of this file is to answer "how much
 * faster could the client be" without giving up determinism.
 *
 * DETERMINISM BY FAITHFUL TRANSCRIPTION: WebAssembly f64 arithmetic
 * (+ - * / sqrt, floor) is IEEE-754 correctly rounded, exactly like
 * JavaScript's. Every routine below performs the SAME operations in
 * the SAME order as its neural.js counterpart, so results are
 * bit-identical - verified against every pinned vector by
 * wasm/wasm.test.mjs and wasm/verify.mjs. Rules observed:
 *
 *  - no libm, no -ffast-math, no reassociation, no FMA (WASM has no
 *    scalar FMA; the SIMD mul/add used below are separate, exactly
 *    rounded instructions - NEVER replace them with relaxed-simd);
 *  - detExp/detLog2 equivalents are transcribed term by term;
 *  - the v3 matmul4 kernel maps its four interleaved accumulators
 *    s0..s3 onto two f64x2 vector accumulators (lanes s0,s1 and
 *    s2,s3), four elements per iteration, combined (s0+s1)+(s2+s3) -
 *    the identical summation order, just two sums per instruction;
 *  - v1/v2 models keep the sequential single-accumulator kernel,
 *    which cannot be vectorized without reordering, so it stays
 *    scalar.
 *
 * Memory management is deliberately trivial: the JS glue (engine.js)
 * owns a bump allocator above __heap_base and passes byte offsets in.
 * Weights arrive as f32 (exactly as neural.js stores them - the f32
 * values are exact, math happens in f64), key/value cache entries are
 * f64 slabs, and per-session tables of k/v offsets let forked
 * sessions share cached positions by pointer exactly like the JS
 * engine's write-once arrays.
 */

#include <wasm_simd128.h>

#define MAX_DIM 1024
#define MAX_MLP 4096
#define MAX_VOCAB 4096
#define MAX_LEN 512
#define MAX_LAYERS 16

/* Exact double constants, identical literals to neural.js */
static const double LN2 = 0.6931471805599453;
static const double INV_LN2 = 1.4426950408889634;

/* POW2[i] = 2^(-i), built by halving (exact for normal doubles) */
static double POW2[1101];

/* Model descriptor, filled by init() */
static int g_dim, g_heads, g_layers, g_mlp, g_vocab, g_maxlen, g_use4;
static double g_attn_scale;

/* Tensor locations (byte offsets into linear memory, set by JS) */
static const float *t_embed, *t_pos, *t_norm;
static const float *t_norm1[MAX_LAYERS], *t_qkv[MAX_LAYERS],
  *t_proj[MAX_LAYERS], *t_norm2[MAX_LAYERS], *t_up[MAX_LAYERS],
  *t_down[MAX_LAYERS];

/* Scratch buffers (one inference step, mirrors the JS session's) */
static double s_x[MAX_DIM], s_h[MAX_DIM], s_qkv[3 * MAX_DIM],
  s_attn[MAX_DIM], s_proj[MAX_DIM], s_mlp[MAX_MLP],
  s_logits[MAX_VOCAB], s_scores[MAX_LEN];

/**
 * Deterministic exp() for non-positive arguments - transcribed from
 * neural.js detExp: reduction by ln2, then a fixed-order 13-term
 * Taylor series. floor lowers to the correctly-rounded f64.floor.
 */
static double det_exp (double x) {
  if (x < -708) return 0;
  double n = __builtin_floor(x * INV_LN2 + 0.5);
  double r = x - n * LN2;
  double term = 1;
  double sum = 1;
  for (int i = 1; i <= 13; i ++) {
    term = term * r / i;
    sum += term;
  }
  return POW2[-(int)n] * sum;
}

/** In-place softmax over x[0..n), transcribed from neural.js. */
static void softmax_n (double *x, int n) {
  double max = -__builtin_inf();
  for (int i = 0; i < n; i ++) {
    if (x[i] > max) max = x[i];
  }
  double sum = 0;
  for (int i = 0; i < n; i ++) {
    x[i] = det_exp(x[i] - max);
    sum += x[i];
  }
  for (int i = 0; i < n; i ++) x[i] /= sum;
}

/** RMS-norm of x scaled by g into out, transcribed from neural.js. */
static void rms_norm (const double *x, const float *g, double *out, int n) {
  double sum = 0;
  for (int i = 0; i < n; i ++) sum += x[i] * x[i];
  double scale = 1 / __builtin_sqrt(sum / n + 1e-5);
  for (int i = 0; i < n; i ++) out[i] = x[i] * scale * (double)g[i];
}

/**
 * Sequential matmul (v1/v2 kernel): one accumulator, strictly
 * left-to-right. The dependency chain forbids vectorization - the
 * summation order is frozen by issued v1/v2 links.
 */
static void matmul_seq (const float *w, int rows, int cols,
    const double *x, double *out) {
  for (int r = 0; r < rows; r ++) {
    double sum = 0;
    const float *base = w + (long)r * cols;
    for (int c = 0; c < cols; c ++) {
      sum += (double)base[c] * x[c];
    }
    out[r] = sum;
  }
}

/**
 * Unrolled matmul (v3 kernel): four interleaved partial sums combined
 * as (s0+s1)+(s2+s3), exactly neural.js matmul4. The two f64x2
 * accumulators hold lanes (s0,s1) and (s2,s3); each lane performs the
 * identical sequence of correctly-rounded mul/add operations the JS
 * scalars do, so the result is bit-identical while running four
 * MACs per iteration.
 */
static void matmul_4 (const float *w, int rows, int cols,
    const double *x, double *out) {
  for (int r = 0; r < rows; r ++) {
    const float *base = w + (long)r * cols;
    v128_t acc01 = wasm_f64x2_splat(0);
    v128_t acc23 = wasm_f64x2_splat(0);
    for (int c = 0; c < cols; c += 4) {
      v128_t wf = wasm_v128_load(base + c);
      v128_t w01 = wasm_f64x2_promote_low_f32x4(wf);
      v128_t w23 = wasm_f64x2_promote_low_f32x4(
        wasm_i32x4_shuffle(wf, wf, 2, 3, 2, 3));
      acc01 = wasm_f64x2_add(acc01, wasm_f64x2_mul(w01,
        wasm_v128_load(x + c)));
      acc23 = wasm_f64x2_add(acc23, wasm_f64x2_mul(w23,
        wasm_v128_load(x + c + 2)));
    }
    out[r] = (wasm_f64x2_extract_lane(acc01, 0)
        + wasm_f64x2_extract_lane(acc01, 1))
      + (wasm_f64x2_extract_lane(acc23, 0)
        + wasm_f64x2_extract_lane(acc23, 1));
  }
}

static void mm (const float *w, int rows, int cols,
    const double *x, double *out) {
  if (g_use4) matmul_4(w, rows, cols, x, out);
  else matmul_seq(w, rows, cols, x, out);
}

/**
 * Configures the engine for a loaded model. use4 selects the v3
 * kernel; v1/v2 models must pass 0 (their payloads froze the
 * sequential summation order).
 */
__attribute__((export_name("init")))
void init (int dim, int heads, int layers, int mlp, int vocab,
    int maxlen, int use4) {
  g_dim = dim;
  g_heads = heads;
  g_layers = layers;
  g_mlp = mlp;
  g_vocab = vocab;
  g_maxlen = maxlen;
  g_use4 = use4;
  g_attn_scale = 1 / __builtin_sqrt((double)(dim / heads));
  POW2[0] = 1;
  for (int i = 1; i < 1101; i ++) POW2[i] = POW2[i - 1] * 0.5;
}

/**
 * Registers a tensor's location. kind: 0 embed, 1 pos, 2 final norm,
 * 3 norm1, 4 qkv, 5 proj, 6 norm2, 7 up, 8 down (3..8 per layer).
 */
__attribute__((export_name("set_tensor")))
void set_tensor (int kind, int layer, unsigned offset) {
  const float *p = (const float *)offset;
  switch (kind) {
    case 0: t_embed = p; break;
    case 1: t_pos = p; break;
    case 2: t_norm = p; break;
    case 3: t_norm1[layer] = p; break;
    case 4: t_qkv[layer] = p; break;
    case 5: t_proj[layer] = p; break;
    case 6: t_norm2[layer] = p; break;
    case 7: t_up[layer] = p; break;
    case 8: t_down[layer] = p; break;
  }
}

/** Address of the probability output buffer (vocab doubles). */
__attribute__((export_name("probs_ptr")))
unsigned probs_ptr (void) {
  return (unsigned)s_logits;
}

/**
 * One transformer step, transcribed from the JS session's feed().
 * The table holds, per layer and cached position, the byte offsets
 * of that position's key and value vectors (dim doubles each):
 * table[(l * maxLen + j) * 2] = key, + 1 = value. The entries for
 * `position` point at freshly allocated slabs this call fills in -
 * the JS glue writes them before calling. After the call the probs
 * buffer holds softmax(logits), which is what both the coder and the
 * tokenization search consume.
 */
__attribute__((export_name("feed")))
void feed (int id, int position, unsigned table_off) {
  const unsigned *table = (const unsigned *)table_off;
  const int dim = g_dim;
  const int head_dim = dim / g_heads;
  const int count = position + 1;

  for (int i = 0; i < dim; i ++) {
    s_x[i] = (double)t_embed[id * dim + i] + (double)t_pos[position * dim + i];
  }

  for (int l = 0; l < g_layers; l ++) {
    // Attention
    rms_norm(s_x, t_norm1[l], s_h, dim);
    mm(t_qkv[l], 3 * dim, dim, s_h, s_qkv);
    const unsigned *row = table + (long)l * g_maxlen * 2;
    double *k_new = (double *)row[position * 2];
    double *v_new = (double *)row[position * 2 + 1];
    for (int i = 0; i < dim; i ++) {
      k_new[i] = s_qkv[dim + i];
      v_new[i] = s_qkv[2 * dim + i];
    }
    for (int head = 0; head < g_heads; head ++) {
      const int base = head * head_dim;
      for (int j = 0; j < count; j ++) {
        double sum = 0;
        const double *k = (const double *)row[j * 2];
        for (int i = 0; i < head_dim; i ++) {
          sum += s_qkv[base + i] * k[base + i];
        }
        s_scores[j] = sum * g_attn_scale;
      }
      softmax_n(s_scores, count);
      for (int i = 0; i < head_dim; i ++) s_attn[base + i] = 0;
      for (int j = 0; j < count; j ++) {
        const double *v = (const double *)row[j * 2 + 1];
        const double weight = s_scores[j];
        for (int i = 0; i < head_dim; i ++) {
          s_attn[base + i] += weight * v[base + i];
        }
      }
    }
    mm(t_proj[l], dim, dim, s_attn, s_proj);
    for (int i = 0; i < dim; i ++) s_x[i] += s_proj[i];

    // MLP
    rms_norm(s_x, t_norm2[l], s_h, dim);
    mm(t_up[l], g_mlp, dim, s_h, s_mlp);
    for (int i = 0; i < g_mlp; i ++) {
      if (s_mlp[i] < 0) s_mlp[i] = 0;
    }
    mm(t_down[l], dim, g_mlp, s_mlp, s_proj);
    for (int i = 0; i < dim; i ++) s_x[i] += s_proj[i];
  }

  rms_norm(s_x, t_norm, s_h, dim);
  // Output head is tied to the embedding table
  mm(t_embed, g_vocab, dim, s_h, s_logits);
  softmax_n(s_logits, g_vocab);
}

File: wasm/engine.js
/**
 * @file JS glue for the production WebAssembly inference engine.
 *
 * Pairs engine.wasm (built from engine.c, recipe in wasm/README.md)
 * with a faithful port of the neural.js encode/decode pipeline.
 * neural.js is untouched and remains the reference implementation and
 * automatic fallback (see engine-select.js at the repo root, which
 * every production entry point goes through): this module re-uses its
 * model loader (URLModel - dequantization, tokenizer, payload
 * versioning) and the shared arithmetic coder, and swaps ONLY the
 * inference engine underneath. Every payload produced or consumed
 * here must be bit-identical to neural.js - that is the acceptance
 * criterion, not a hope (see wasm/wasm.test.mjs and wasm/verify.mjs).
 *
 * Memory model: the WASM module owns its linear memory. At load, the
 * dequantized f32 tensors are copied in once. Above the weights lives
 * a bump-allocated arena for per-session key/value slabs and the
 * per-session tables of k/v offsets that let forked sessions share
 * cached positions by pointer, exactly like the JS engine's
 * write-once per-position arrays. reset() reclaims the whole arena;
 * callers only invoke it when no session is live (between candidate
 * encodes, and at each chunk restart, where the previous chunk's
 * session is dead by construction).
 */

import { URLModel } from "../neural.js";
import { arithmeticEncode, arithmeticDecode } from "../arithmetic-coder.js";

const EOS = 0;

/** Exact double constant, same literal as neural.js */
const INV_LN2 = 1.4426950408889634;

/**
 * Deterministic base-2 logarithm, transcribed from neural.js detLog2
 * (not exported there). Only used for the encoder's search costs; the
 * source is identical, so results are bit-identical.
 * @param {number} x Argument, must be >= 0 and finite
 * @returns {number} log2(x), or -Infinity for zero
 */
function detLog2 (x) {
  if (x === 0) return -Infinity;
  let e = 0;
  while (x < 1) { x *= 2; e --; }
  while (x >= 2) { x *= 0.5; e ++; }
  const z = (x - 1) / (x + 1);
  const z2 = z * z;
  let term = z;
  let sum = z;
  for (let i = 3; i <= 29; i += 2) {
    term *= z2;
    sum += term / i;
  }
  return e + 2 * sum * INV_LN2;
}

/** Tensor-name -> engine.c set_tensor kind (3..8 are per-layer). */
const TENSOR_KINDS = { norm1: 3, qkv: 4, proj: 5, norm2: 6, up: 7, down: 8 };

const align16 = (n) => (n + 15) & ~15;

/**
 * A loaded model bound to a WASM instance. One instance per engine;
 * weights are copied into linear memory once at construction.
 */
export class WasmEngine {
  /**
   * Instantiates the WASM module and copies the model's weights in.
   * @param {URLModel|ArrayBuffer} model Parsed model (or raw file)
   * @param {BufferSource|WebAssembly.Module} wasmBinary engine.wasm
   * @returns {Promise<WasmEngine>} Ready engine
   */
  static async create (model, wasmBinary) {
    if (!(model instanceof URLModel)) model = new URLModel(model);
    const module = wasmBinary instanceof WebAssembly.Module
      ? wasmBinary : await WebAssembly.compile(wasmBinary);
    const instance = await WebAssembly.instantiate(module, {});
    return new WasmEngine(model, instance);
  }

  /**
   * @param {URLModel} model Parsed model
   * @param {WebAssembly.Instance} instance Instantiated engine.wasm
   */
  constructor (model, instance) {
    this.model = model;
    this.exports = instance.exports;
    this.memory = instance.exports.memory;

    const entries = Object.entries(model.tensors);
    let offset = align16(this.exports.__heap_base.value);
    let total = offset;
    for (const [, data] of entries) total += align16(data.length * 4);
    this.grow(total);

    for (const [name, data] of entries) {
      new Float32Array(this.memory.buffer, offset, data.length).set(data);
      let kind;
      let layer = 0;
      if (name === "embed") kind = 0;
      else if (name === "pos") kind = 1;
      else if (name === "norm") kind = 2;
      else {
        const match = /^b(\d+)\.(\w+)$/.exec(name);
        layer = Number(match[1]);
        kind = TENSOR_KINDS[match[2]];
      }
      this.exports.set_tensor(kind, layer, offset);
      offset += align16(data.length * 4);
    }
    this.exports.init(model.dim, model.heads, model.layers, model.mlpDim,
      model.vocab, model.maxLen, model.fastKernels ? 1 : 0);
    this.probsPtr = this.exports.probs_ptr();
    this.arenaBase = align16(offset);
    this.arenaNext = this.arenaBase;
  }

  /** Grows linear memory so `bytes` total bytes are addressable. */
  grow (bytes) {
    const current = this.memory.buffer.byteLength;
    if (bytes > current) {
      // Grow with 16MB headroom so per-feed allocations don't grow
      // one page at a time
      this.memory.grow(Math.ceil((bytes - current) / 65536) + 256);
    }
  }

  /** Bump-allocates 16-aligned arena bytes; returns the byte offset. */
  alloc (bytes) {
    const ptr = this.arenaNext;
    this.arenaNext = ptr + align16(bytes);
    this.grow(this.arenaNext);
    return ptr;
  }

  /**
   * Reclaims the whole session arena. Only call with no live
   * sessions - every session's tables and k/v slabs live there.
   */
  reset () {
    this.arenaNext = this.arenaBase;
  }

  /**
   * Starts an inference session mirroring URLModel.session(): `feed`
   * runs one transformer step, `fork` clones the session so the
   * tokenization search can branch without re-feeding the shared
   * prefix (the k/v slabs are shared by table copy - written once,
   * read-only afterwards). Unlike the JS engine, `feed` returns the
   * softmaxed probabilities (both of neural.js's consumers softmax
   * the returned logits immediately, so this is the same math).
   * @returns {{feed: (id: number) => Float64Array, fork: () => object}}
   */
  session () {
    const { layers, maxLen, dim, vocab } = this.model;
    const tableBytes = layers * maxLen * 2 * 4;
    const engine = this;

    const spawn = (srcTable, positionInit) => {
      const table = engine.alloc(tableBytes);
      if (srcTable !== null) {
        new Uint8Array(engine.memory.buffer)
          .copyWithin(table, srcTable, srcTable + tableBytes);
      }
      let position = positionInit;

      const feed = (id) => {
        if (position >= maxLen) throw "Model context window exceeded.";
        const kv = engine.alloc(layers * 2 * dim * 8);
        const entries =
          new Uint32Array(engine.memory.buffer, table, layers * maxLen * 2);
        for (let l = 0; l < layers; l ++) {
          entries[(l * maxLen + position) * 2] = kv + l * 2 * dim * 8;
          entries[(l * maxLen + position) * 2 + 1] = kv + (l * 2 + 1) * dim * 8;
        }
        engine.exports.feed(id, position, table);
        position ++;
        return new Float64Array(
          engine.memory.buffer, engine.probsPtr, vocab).slice();
      };

      const fork = () => spawn(table, position);

      return { feed, fork };
    };

    return spawn(null, 0);
  }
}

/* ---- Pipeline ported from neural.js, engine swapped underneath ---- */

function chunkCapacity (model) {
  return model.maxLen - 2;
}

const CHUNKED_MAX_SYMBOLS = 4096;

/** Chunk framing, transcribed from neural.js chunkFrame. */
function chunkFrame (model, symbols) {
  const capacity = chunkCapacity(model);
  const framed = [];
  for (let i = 0; i < symbols.length; i += capacity) {
    framed.push(...symbols.slice(i, i + capacity), EOS);
  }
  if (symbols.length % capacity === 0) {
    framed.push(EOS);
  }
  return framed.length > CHUNKED_MAX_SYMBOLS ? null : framed;
}

/**
 * Probability callback for the arithmetic coder, transcribed from
 * neural.js modelProbabilities with a WASM session underneath. At a
 * chunk restart the previous chunk's session is dead, so the arena is
 * reclaimed before spawning the fresh session.
 * @param {WasmEngine} engine Loaded engine
 * @param {{context: number[]}} [chunkState] Latest-context out-param
 * @returns {(context: number[]) => Float64Array} Probability callback
 */
export function wasmProbabilities (engine, chunkState = null) {
  const chunked = engine.model.linkVersion >= 3;
  let session = engine.session();
  let fed = 0;
  return (context) => {
    let probs = null;
    if (fed === 0) {
      probs = session.feed(EOS);
      fed = 1;
    }
    while (fed <= context.length) {
      const symbol = context[fed - 1];
      if (chunked && symbol === EOS) {
        engine.reset();
        session = engine.session();
        probs = session.feed(EOS);
      } else {
        probs = session.feed(symbol);
      }
      fed ++;
    }
    if (probs === null) {
      throw "Arithmetic coder context out of sync with model session.";
    }
    if (chunkState) chunkState.context = context;
    return probs;
  };
}

/** Trailing-run length rule, transcribed from neural.js. */
function closedChunkLength (context) {
  let i = context.length - 1;
  if (i >= 0 && context[i] === EOS) i --;
  let run = 0;
  for (; i >= 0 && context[i] !== EOS; i --) {
    run ++;
  }
  return run;
}

const BEAM_WIDTH = 4;
const FINAL_CANDIDATES = 2;

/**
 * Encoder-side tokenization search, transcribed from neural.js
 * searchSegmentations. The only change: WASM sessions already return
 * softmaxed probabilities, so the explicit softmax step is gone (the
 * numbers are identical). Costs use the same detLog2, hypotheses
 * compare identically, so the chosen candidates - and therefore the
 * payload - match the JS encoder bit for bit.
 * @param {WasmEngine} engine Loaded engine (model has a token list)
 * @param {string} text Scheme-less URL text
 * @returns {number[][]} Candidate symbol sequences (without EOS)
 */
function searchSegmentations (engine, text) {
  const model = engine.model;
  const maxSymbols = model.maxLen - 2;
  const length = text.length;

  const compare = (a, b) => {
    if (a.cost !== b.cost) return a.cost < b.cost ? -1 : 1;
    const n = Math.min(a.symbols.length, b.symbols.length);
    for (let i = 0; i < n; i ++) {
      if (a.symbols[i] !== b.symbols[i]) return a.symbols[i] - b.symbols[i];
    }
    return a.symbols.length - b.symbols.length;
  };

  const arrivals = [];
  for (let p = 0; p <= length; p ++) arrivals.push([]);
  const finals = [];

  const expand = (hyp, p) => {
    if (hyp.symbols.length >= maxSymbols) return;
    const limit = Math.min(model.maxTokenLength, length - p);
    for (let l = 1; l <= limit; l ++) {
      const id = model.tokenIds.get(text.slice(p, p + l));
      if (id === undefined) continue;
      arrivals[p + l].push({
        parent: hyp,
        id,
        cost: hyp.cost - detLog2(hyp.probs[id]),
        symbols: hyp.symbols.concat(id)
      });
    }
  };

  const root = { session: engine.session(), symbols: [], cost: 0 };
  root.probs = root.session.feed(EOS);
  expand(root, 0);

  for (let p = 1; p <= length; p ++) {
    const candidates = arrivals[p];
    if (!candidates.length) continue;
    candidates.sort(compare);
    for (const candidate of candidates.slice(0, BEAM_WIDTH)) {
      const session = candidate.parent.session.fork();
      const probs = session.feed(candidate.id);
      if (p < length) {
        expand({ session, symbols: candidate.symbols,
          cost: candidate.cost, probs }, p);
      } else {
        finals.push({ symbols: candidate.symbols,
          cost: candidate.cost - detLog2(probs[EOS]) });
      }
    }
  }

  finals.sort(compare);
  return finals.slice(0, FINAL_CANDIDATES).map(final => final.symbols);
}

/**
 * Arithmetic-codes one symbol sequence into a framed payload number,
 * transcribed from neural.js encodeSymbols. Resets the arena first:
 * any sessions from a preceding search or candidate encode are dead.
 * @param {WasmEngine} engine Loaded engine
 * @param {number[]} symbols Symbol ids (without EOS)
 * @param {boolean} isHTTPS Whether the link scheme is https
 * @returns {BigInt?} Payload number
 */
function encodeSymbols (engine, symbols, isHTTPS) {
  const model = engine.model;
  const coded = model.linkVersion >= 3
    ? chunkFrame(model, symbols)
    : symbols.concat(EOS);
  if (coded === null) return null;
  engine.reset();
  const bits = arithmeticEncode(coded, wasmProbabilities(engine));

  let number = 1n;
  for (const bit of bits) {
    number = (number << 1n) | BigInt(bit);
  }
  number = (number << 1n) | (isHTTPS ? 1n : 0n);
  const version = BigInt(model.linkVersion);
  number = (number << (version + 1n)) | ((1n << version) - 1n);
  return number;
}

/**
 * Compresses a link into a raw payload number - the WASM-backed
 * counterpart of neural.js neuralCompressToNumber, same options,
 * bit-identical output.
 * @param {WasmEngine} engine Loaded engine
 * @param {string} input Link to compress
 * @param {{search?: boolean}} [options] `search: false` = greedy only
 * @returns {BigInt?} Payload number, or null if the link doesn't fit
 */
export function wasmCompressToNumber (engine, input, { search = true } = {}) {
  const model = engine.model;
  let url;
  if (URL.canParse(input)) {
    url = new URL(input);
  } else {
    url = new URL("http://" + input);
  }
  if (url.protocol !== "http:" && url.protocol !== "https:") return null;
  const isHTTPS = url.protocol === "https:";
  const text = url.href.slice(isHTTPS ? 8 : 7);

  const greedy = model.tokenize(text);
  if (greedy === null) return null;
  if (model.linkVersion < 3 && greedy.length + 2 > model.maxLen) return null;

  const candidates = [greedy];
  if (search && model.tokens && model.linkVersion >= 2 &&
      greedy.length + 2 <= model.maxLen) {
    engine.reset();
    candidates.push(...searchSegmentations(engine, text));
  }

  let best = null;
  const seen = new Set();
  for (const symbols of candidates) {
    const key = symbols.join(",");
    if (seen.has(key)) continue;
    seen.add(key);
    const number = encodeSymbols(engine, symbols, isHTTPS);
    if (number === null) continue;
    if (best === null || number < best) best = number;
  }
  return best;
}

/**
 * Decompresses a payload number back into a full link - the
 * WASM-backed counterpart of neural.js neuralDecompressNumber.
 * @param {WasmEngine} engine Loaded engine
 * @param {BigInt} number Payload number (version marker included)
 * @returns {string} Full link
 */
export function wasmDecompressNumber (engine, number) {
  const model = engine.model;
  let version = 0;
  while (number & 1n) {
    version ++;
    number >>= 1n;
  }
  number >>= 1n;
  if (version !== model.linkVersion) {
    throw `Payload version ${version} needs model version ${version}, `
      + `but this model is version ${model.linkVersion}.`;
  }
  const isHTTPS = number & 1n;
  number >>= 1n;

  const bitString = number.toString(2);
  if (bitString === "0") throw "Empty neural payload.";
  const bits = [];
  for (let i = 1; i < bitString.length; i ++) {
    bits.push(bitString.charCodeAt(i) - 0x30);
  }

  engine.reset();
  let symbols;
  if (model.linkVersion >= 3) {
    const capacity = chunkCapacity(model);
    const chunkState = { context: [] };
    symbols = arithmeticDecode(
      bits, wasmProbabilities(engine, chunkState),
      (s) => s === EOS && closedChunkLength(chunkState.context) < capacity,
      CHUNKED_MAX_SYMBOLS);
    symbols = symbols.filter((s) => s !== EOS);
  } else {
    symbols = arithmeticDecode(
      bits, wasmProbabilities(engine), (s) => s === EOS, model.maxLen);
    symbols.pop();
  }
  return (isHTTPS ? "https://" : "http://") + model.detokenize(symbols);
}
